lib/pbench/agent/results.py

```diff
@@ -14,45 +14,79 @@
 from pbench.common.utils import md5sum, validate_hostname
 
 
-class FileUploadError(Exception):
-    """Raised when the uploading of file to server has failed"""
-
-    pass
-
-
 class MakeResultTb:
     """MakeResultTb - Creates the result tar file.
 
-        TODO:  This is latent code -- it is currently unused and largely
-               untested, intended to support the future implementation of
-               a tool to replace pbench-make-result-tb.
     """
 
+    class AlreadyCopied(RuntimeError):
+        """Specific run time error raised when a result directory has a .copied
+        marker present.
+        """
+
+        pass
+
+    class BenchmarkRunning(RuntimeError):
+        """Specific run time error raised when a result directory has a .running
+        marker present.
+        """
+
+        pass
+
     def __init__(
         self,
         result_dir: str,
         target_dir: str,
+        controller: str,
         config: PbenchAgentConfig,
         logger: Logger,
     ):
-        """__init__ - Initializes the required attributes
+        """__init__ - Initializes the required attributes for making a tar ball
 
-            Args:
-                result_dir -- directory where tb file is collected.
-                target_dir -- directory where tar file needs to be moved.
-                config -- PbenchAgent config object
-                logger -- logger objects helps logging important details
+        Args:
+            result_dir -- directory where tb file is collected.
+            target_dir -- directory where tar file needs to be moved.
+            config -- PbenchAgent config object
+            logger -- logger objects helps logging important details
+
+        Raises:
+            AlreadyCopied       if we find a "*.copied" marker for the result
+                                directory.
+            BenchmarkRunning    if we find a "*/.running" marker in the result
+                                directory.
+            FileNotFoundError   if either the result or target directories do
+                                not exist
+            NotADirectoryError  if either the result or target directories are
+                                not actual directories
+            RuntimeError        if it cannot find the 'tar' command on the PATH
         """
         assert (
             config and logger
         ), f"config, '{config!r}', and/or logger, '{logger!r}', not provided"
-        self.result_dir = self.check_result_target_dir(result_dir, "Result")
-        self.target_dir = self.check_result_target_dir(target_dir, "Target")
+        self.tar_path = find_executable("tar")
+        if self.tar_path is None:
+            raise RuntimeError(f"External 'tar' executable not found")
+        self.xz_path = find_executable("xz")
+        if self.xz_path is None:
+            raise RuntimeError(f"External 'xz' executable not found")
```

Comment:
The problem is we don't want the defined a class definition time (when the module is imported), but once the "main" program has started running.

Reply:
What do you think is going to change or happen between "class definition time" and when the "main" program starts running?

My understanding is that Python code is just a script which gets read from top to bottom executing each line in sequence -- things like `class` statements have effects when they are read -- and it's only when a class or function gets referenced (or invoked) that stuff "runs" as we think of it.  Import statements are just another thing which gets executed, as are initializers for global variables and class attributes.  So, yeah, moving these calls of `find_executable()` from instance initializers to class initializers would mean that they get executed at a different time, but that's exactly what I want -- they'll get executed only once overall instead of once for each instance.  Are you saying that the current working directory or the shell path (or something else which will affect their operation) will be different after the main function gets called?

Reply:
> Are you saying that the current working directory or the shell path (or something else which will affect their operation) will be different after the main function gets called?

Sort of.

When `test_results_move` imports the `move` module, that intern imports the `MakeResultTb` class.  So just by loading the test, the `find_executable()` methods would be invoked at module load time.  This makes it hard to mock out the `tar` and `xz` commands.

In the case of the Tool Meister code, was have everything in one module, so we can do it once on when the command actually executes.

But here, we either have to pass in the command paths to use each time, or make the optional parameters to override the class definitions.  Hmm.

Reply:
Ah, yes...I hadn't considered a requirement to be able to mock the `find_executable()` call.

Of course, if these attributes were class attributes, the test code could simply reassign them without having to patch or mock anything.

Reply:
Unit tests can certainly solve that by slamming the class variables to what they require, but that prevents other kind of testing.

Reply:
Outside of unit testing, I think having the values calculated only once (and stored as class attributes) is a better solution (because it avoids unnecessary time spent in instance initialization).  Now, this probably isn't performance-sensitive code, and, in most scenarios, we probably never create more than one instance of the class, and so we probably don't save much by doing the initialization during class initialization instead of during the (one) object instantiation, but that's not an argument _against_ using class initialization.

As for unit testing, being able to write the class attributes is probably exactly equivalent to monkeypatching class member functions with mocks, and it's simpler and probably clearer.

So, what other testing did you have in mind that using class attributes would interfere with?

Reply:
Where is the unit test hooking the `tar` and `xz` programs?  (I can't find it in the source...is it doing it via the shell's `$PATH`?)

Reply:
Other kinds of testing where you don't want to have to actually run the `tar` and `xz` commands, but mock out their behavior like we do in the legacy unit tests.

We have no other examples of this in the Tool Meister code where those commands are found at run time so that they can be externally overridden.

Reply:
For unit testing, we shouldn't actually be running `tar`/`xz` or even external proxies for them.  We should be mocking `find_executable()` and `subprocess.Popen()`.  And, therefore, we don't care how these external functions work because we will control how the CUT sees them behave.

As for functional testing, we should be running the real things, and we shouldn't be concerned with trying to run proxies in their stead.

So, I have very little enthusiasm for setting up any kind of mid-ground, especially if the justification is that we're following the legacy testing model which we would prefer to be getting away from.

Reply:
Let's continue this discussion in PR #2443.